samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -215,14 +242,60 @@ public void close() throws ZkInterruptedException {
     zkClient.close();
   }
 
+  // Generation enforcing zk listener abstract class.
+  // Helps listeners, which extend it,  to skip old generation events.
+  // We cannot use 'sessionId' for this because it is not available through ZkClient (at leaste without reflection)
+  public abstract static class GenIZkChildListener implements IZkChildListener {
+    private final int generation;
+    private final ZkUtils zkUtils;
+
+    public GenIZkChildListener(ZkUtils zkUtils) {
+      generation = zkUtils.getGeneration();
+      this.zkUtils = zkUtils;
+    }
+
+    protected boolean skip(String listenerName) {
+      int curGeneration = zkUtils.getGeneration();
+      if (curGeneration != generation) {
+        LOG.warn("SKIPPING handleDataChanged for " + listenerName +
+            " from wrong generation. curGen=" + curGeneration + "; cb gen= " + generation);
+        return true;
+      }
+      return false;
+    }
+  }
+
+  public abstract static class GenIZkDataListener implements IZkDataListener {
+    private final int generation;
+    private final ZkUtils zkUtils;
+
+    public GenIZkDataListener(ZkUtils zkUtils) {
+      generation = zkUtils.getGeneration();
+      this.zkUtils = zkUtils;
+    }
+
+    protected boolean skip(String listenerName) {
+      int curGeneration = zkUtils.getGeneration();
+      if (curGeneration != generation) {
+        LOG.warn("SKIPPING handleDataChanged for " + listenerName +
+            " from wrong generation. curGen=" + curGeneration + "; cb gen= " + generation);
+        return true;
+      }
+      return false;
+    }
+
+  }
+
   /**
     * subscribe for changes of JobModel version
     * @param dataListener describe this
     */
-  public void subscribeToJobModelVersionChange(IZkDataListener dataListener) {
+  public void subscribeToJobModelVersionChange(GenIZkDataListener dataListener) {
     LOG.info(" subscribing for jm version change at:" + keyBuilder.getJobModelVersionPath());
     zkClient.subscribeDataChanges(keyBuilder.getJobModelVersionPath(), dataListener);
-    metrics.subscriptions.inc();
+    if (metrics != null) {
```

Comment:
It's initialized with proper params (will be non null). 

Even if we want to guard against null, ll be great to verify once in constructor before assignment(just to eliminate repeated checks).